leshan-tl-javacoap-server/src/main/java/org/eclipse/leshan/transport/javacoap/endpoint/JavaCoapServerEndpoint.java

```diff
@@ -0,0 +1,233 @@
+        this.toolbox = toolbox;
+        this.notificationReceiver = notificationReceiver;
+        this.registrationStore = registrationStore;
+
+    }
+
+    @Override
+    public Protocol getProtocol() {
+        return Protocol.COAP;
+    }
+
+    @Override
+    public URI getURI() {
+        return endpointUri;
+    }
+
+    @Override
+    public <T extends LwM2mResponse> T send(ClientProfile destination, DownlinkRequest<T> request,
+            LowerLayerConfig lowerLayerConfig, long timeoutInMs) throws InterruptedException {
+
+        // Send LWM2M Request
+        CompletableFuture<T> lwm2mResponseFuture = sendLwM2mRequest(destination, request, lowerLayerConfig);
+
+        // Wait synchronously for LWM2M response
+        try {
+            return lwm2mResponseFuture.get(timeoutInMs, TimeUnit.MILLISECONDS);
+        } catch (CompletionException | ExecutionException exception) {
+            // TODO we probably need to enhance this (better translate java-coap exceptions to leshan ones)
+            throw new SendFailedException("Unable to send request", exception);
+        } catch (TimeoutException e) {
+            lwm2mResponseFuture.cancel(true);
+            return null;
+        }
+    }
+
+    @Override
+    public <T extends LwM2mResponse> void send(ClientProfile destination, DownlinkRequest<T> request,
+            ResponseCallback<T> responseCallback, ErrorCallback errorCallback, LowerLayerConfig lowerLayerConfig,
+            long timeoutInMs) {
+
+        // Send LWM2M Request
+        CompletableFuture<T> lwm2mResponseFuture = sendLwM2mRequest(destination, request, lowerLayerConfig);
+
+        // Attach callback
+        lwm2mResponseFuture
+                // Handle Exception
+                .exceptionally((exception) -> {
+                    // TODO we probably need to enhance this (better translate java-coap exceptions to leshan ones)
+                    errorCallback.onError(new SendFailedException("Unable to send request", exception));
+                    return null;
+                })
+                // Handle CoAP Response
+                .thenAccept((lwM2mResponse) -> {
+                    responseCallback.onResponse(lwM2mResponse);
+                });
+
+        // TODO handle timeout (cancel future)
+    }
+
+    protected <T extends LwM2mResponse> CompletableFuture<T> sendLwM2mRequest(ClientProfile destination,
+            DownlinkRequest<T> lwm2mRequest, LowerLayerConfig lowerLayerConfig) {
+
+        if (lwm2mRequest instanceof ObserveRequest) {
+            return sendObserveRequest(destination, lwm2mRequest, lowerLayerConfig);
+        } else {
+            // Create Coap Request to send from LWM2M Request
+            CoapRequest coapRequest = translator.createCoapRequest(destination, lwm2mRequest, toolbox);
+
+            // Create a Coap Client to send request
+            CoapClient coapClient = CoapClientBuilder.clientFor(destination.getIdentity().getPeerAddress(), coapServer);
```

Comment:
You do not need to create coapclient just to send request. 
You can use `coapServerserver.clientService().apply(..)`

Reply:
OK I will change this.

But if I should not use it, what is the use case for `CoapClientBuilder.clientFor(destination, coapServer)`

Reply:
It's just a helper wrapper. It will place destination address in every request so no need to define it. Mostly useful in tests.

Reply:
Ok done :)